lang/LangSource/PyrSlot.h

```diff
@@ -27,16 +27,398 @@ A PyrSlot is an 8-byte value which is either a double precision float or a
+//      | false             | true
+//  * double *              |
+//                          |
+//                          |
+//                *
+
+
+static constexpr uint64_t boxedBit = 0x0001000000000000;
+static constexpr uint64_t nanExponent = 0x7FF8000000000000;
+static constexpr uint64_t boxed = boxedBit | nanExponent;
+
+static constexpr uint64_t pointerMask = 0x0000FFFFFFFFFFFF;
+
+static constexpr uint64_t objHdrTag = 0xFFFF000000000000; // boxed | 0x0001000000000000;
+static constexpr uint64_t intTag = 0x7FFB000000000000; //   boxed | 0x0003000000000000;
+static constexpr uint64_t symTag = 0x7FFD000000000000; // boxed | 0x0005000000000000;
+static constexpr uint64_t charTag = 0x7FFF000000000000; // boxed | 0x0007000000000000;
+static constexpr uint64_t nilFalseTrueTag = 0xFFF9000000000000; // boxed | 0x4001000000000000;
+static constexpr uint64_t ptrTag = 0xFFFB000000000000; // boxed | 0x4003000000000000;
+
+static_assert(objHdrTag != intTag);
+static_assert(objHdrTag != symTag);
+static_assert(objHdrTag != charTag);
+static_assert(objHdrTag != ptrTag);
+
+static constexpr uint64_t nilValue = nilFalseTrueTag | 0x0000000000000000;
+static constexpr uint64_t falseValue = nilFalseTrueTag | 0x0000000000000001;
+static constexpr uint64_t trueValue = nilFalseTrueTag | 0x0000000000000002;
+
+template <typename T> struct PaddedNumber {
+    static_assert(sizeof(T) <= 4);
+    static_assert(std::is_arithmetic_v<T>);
+    static constexpr size_t numExtraBytes = 4 - sizeof(T);
+
+    constexpr explicit PaddedNumber(T v) noexcept: value(v) {}
+    constexpr PaddedNumber() = default;
+
+    ~PaddedNumber() noexcept = default;
+
+#if (BYTE_ORDER == LITTLE_ENDIAN) // TODO: this could be the wrong way around!
+    T value;
+    uint8_t paddingForValue[sizeof(uint32_t) + numExtraBytes] = { 0 };
+#else
+    uint8_t paddingForValue[sizeof(uint32_t) + numExtraBytes] = { 0 };
+    T value;
+#endif
+};
+
+static_assert(sizeof(PaddedNumber<uint32_t>) == 8);
+static_assert(sizeof(PaddedNumber<uint16_t>) == 8);
+static_assert(sizeof(PaddedNumber<uint8_t>) == 8);
+
+template <typename T> struct MaybePaddedPointer {
+    static_assert(std::is_pointer_v<T>);
+#if POINTER_NEEDS_PADDING && (BYTE_ORDER == BIG_ENDIAN)
+    int32_t padding;
+#endif
+
+    T ptr;
+
+    [[nodiscard]] T getPtr() const noexcept {
+#if POINTER_NEEDS_PADDING
+        return ptr; // If there is padding, the tag hasn't tainted the pointer, return it.
+#else
+        return reinterpret_cast<T>(reinterpret_cast<uintptr_t>(ptr) & pointerMask);
+#endif
+    }
+
+#if POINTER_NEEDS_PADDING && (BYTE_ORDER == LITTLE_ENDIAN)
+    int32_t padding;
+#endif
+};
+
+static_assert(sizeof(MaybePaddedPointer<void*>) == sizeof(double));
+
+// cpp reference
+template <class To, class From>
+std::enable_if_t<sizeof(To) == sizeof(From) && std::is_trivially_copyable_v<From> && std::is_trivially_copyable_v<To>,
+                 To>
+bit_cast(const From& src) noexcept {
```

Comment:
You're a better judge than I, but it seems like there are many things in this file that are more or less implementation details for `PyrSlot`. Could we move those to a `namespace detail`?